### PR TITLE
Fix ``FullInliner`` not preserving evaluation order of call arguments

### LINKED ISSUE #14399: FullInliner evaluates inlined function arguments in the wrong order
*Originally reported by a user*.

In Yul the function arguments are evaluated right to left. Unfortunately, the code generated by the `FullInliner` optimizer step (`i`) does not preserve that property.

### Repro
`test.sol`:
```solidity
contract C {
    function f() public {
        assembly ("memory-safe") {
            function fun_revert() -> ret { revert(0, 0) }
            function fun_return() -> ret { return(0, 0) }
            function empty(a, b) {}

            empty(fun_return(), fun_revert())
        }
    }
}
```
```bash
solc test.sol --debug-info none --optimize --yul-optimizations "i" --ir-optimized
```

Without the inliner, the `empty(fun_return(), fun_revert())` call is not inlined and the evaluation order stays the same. This way `fun_revert()` will run first and a call to `f()` will revert.

The inliner, however, works as if `fun_return()` were to be evaluated first, which results in code that `return`s instead:

```yul
if eq(0x26121ff0, newValue)
{
    if callvalue() { revert(0, 0) }
    let headStart := 4
    let dataEnd := calldatasize()
    if slt(sub(dataEnd, headStart), 0) { revert(0, 0) }
    let usr$ret := 0
    return(usr$ret, usr$ret)
}
```

## Environment

- Compiler version: affects at least 0.8.13 and 0.8.20.

> we change the order of visitation, **and** prevent it from inlining when it's anything other than identifiers **and** document the dependency.

### PR DESCRIPTION
**Pull request by cameel**
Fixes #14399.

### CODE REVIEW

File: docs/bugs.json
Comment (cameel) [inline docs/bugs.json]: The UID may have to be adjusted depending on which other important bugfixes get merged first.

File: docs/bugs.json
Comment (cameel) [inline docs/bugs.json]: No blog post yet.

**Comment by ekpyron** [comment]
Once nitpick in the description is: the prerequisite is that the code is in *expression split form*, not that the expression splitter is run before it (or even immediately before it - even the default sequence has steps in between but only those that preserve expression-split form). Might be worthwhile to be clear about that.

File: libyul/optimiser/FullInliner.cpp
Comment (nikola-matic) [inline libyul/optimiser/FullInliner.cpp]:
```suggestion
			return false;
```

File: Changelog.md
Comment (nikola-matic) [inline Changelog.md]: I'm assuming we're not mentioning the case where function arguments are complex expressions (i.e. not identifiers or literals) since that was how it was supposed to work anyway?

File: docs/internals/optimizer.rst
Comment (nikola-matic) [inline docs/internals/optimizer.rst]:
```suggestion
FunctionHoister and ExpressionSplitter are recommended as prerequisites since they make the step
```

File: docs/bugs.json
Comment (nikola-matic) [reply docs/bugs.json]: Is the blog post a prerequisite for this being merged?

File: Changelog.md
Comment (cameel) [reply Changelog.md]: All 3 cases were affected (calls, identifiers and literals), but only the first one could result in wrong bytecode. Whether the latter two worked is a bit of a philosophical distinction :) I did not go into detail here simply to keep this shorter. The inliner was not preserving the order and now it does so it's accurate. The bug list entry and the blog post can go into more detail.

File: docs/internals/optimizer.rst
Comment (cameel) [reply docs/internals/optimizer.rst]: Fixed.

These docs are pretty inconsistent about it though. You'll find all the possible forms (`FunctionHoister`, `Function Hoister` and `function hoister`) all over the place :)

File: Changelog.md
Comment (nikola-matic) [reply Changelog.md]: Yes, I'm aware of that; what I was asking about was the fact that we used to (and still do before this is merged) inline function calls with function call arguments, even though this was never supposed to be the case as per the docs. The major issue here of course was the evaluation order of course.

File: libyul/optimiser/FullInliner.cpp
Comment (cameel) [reply libyul/optimiser/FullInliner.cpp]: Fixed.

File: docs/bugs.json
Comment (cameel) [reply docs/bugs.json]: Not necessarily, but to merge it I'll either have to think what URL will be or omit this for now :)

**Review by nikola-matic — DISMISSED** [review]
I guess it's up to @ekpyron now. Should this still be a draft?

File: Changelog.md
Comment (cameel) [reply Changelog.md]: Ah, that's what you meant. Well technically it's a separate bug so I guess it could have a separate entry. Added one now.

File: docs/bugs.json
Comment (cameel) [reply docs/bugs.json]: Added a tentative link. We still need to remember to recheck that it matches the actual link before the blog post is published.

**Comment by ekpyron** [comment]
On a quick look, the fix looks good, I mainly need to double-check the buglist entry. We should also draft a blog-post about this.

File: Changelog.md
Comment (ekpyron) [inline Changelog.md]: Despite Changelog entries preferably being concise, in this case, should we note that the default optimizer sequence and hence default compiler settings are unaffected by this?

Comment (ekpyron) [reply Changelog.md]: Could be done concisely by just a
```i.e. when using a custom optimizer sequence in which the step is not preceded by ``ExpressionSplitter`` (``x``)```

Comment (cameel) [reply Changelog.md]: Done.

File: docs/bugs.json
Comment (cameel) [reply docs/bugs.json]: #14366 is now using `SOL-2023-1` and I switched to `SOL-2023-2`. Let's just make sure to use subsequent numbers in other PRs (and also that #14366 actually gets merged :).

File: docs/bugs.json
Comment (nikola-matic) [inline docs/bugs.json]:
```suggestion
        "description": "Function call arguments in Yul are evaluated right to left. This order matters when the argument expressions have side-effects, and changing it may change contract behavior. FullInliner is an optimizer step that can replace a function call with the body of that function. The transformation involves assigning argument expressions to temporary variables, which imposes an explicit evaluation order. FullInliner was written with the assumption that this order does not necessarily have to match usual argument evaluation order because the argument expressions have no side-effects. In most circumstances this assumption is true because the default optimization step sequence contains the ExpressionSplitter step. ExpressionSplitter ensures that the code is in *expression-split form*, which means that function calls cannot appear nested inside expressions, and all function call arguments have to be variables. The assumption is, however, not guaranteed to be true in general. Version 0.6.7 introduced a setting allowing users to specify an arbitrary optimization step sequence, making it possible for the FullInliner to actually encounter argument expressions with side-effects, which can result in behavior differences between optimized and unoptimized bytecode. Contracts compiled without optimization or with the default optimization sequence are not affected. To trigger the bug the user has to explicitly choose compiler settings that contain a sequence with FullInliner step not preceded by ExpressionSplitter.",
```

Comment (nikola-matic) [reply docs/bugs.json]: Haha, since it's not that obvious:
```
to explicitly choose compiler settings **than** contain a sequence with FullInliner
```
to
```
to explicitly choose compiler settings **that** contain a sequence with FullInliner
```

**Review by nikola-matic — COMMENTED** [review]
One more typo fix, and this can be merged.

File: docs/bugs.json
Comment (cameel) [reply docs/bugs.json]: Fixed.

**Review by nikola-matic — APPROVED** [review]

**Comment by ekpyron** [comment]
Credit for finding this bug (e.g. for the blog post) goes to "notdeghost"

**Comment by ekpyron** [comment]
> Credit for finding this bug (e.g. for the blog post) goes to "notdeghost"

We just got a follow-up, saying that it should be "Robert Chen from OtterSec" instead.

**Comment by cameel** [comment]
Yeah, saw that. I put that in the blog post: https://github.com/ethereum/solidity-website/pull/11.

**Review by ekpyron — APPROVED** [review]

**Pull request merged**
